File: extensions/aliexpress/src/main/java/app/morphe/extension/aliexpress/patches/RemoveAdsPatch.java
package app.morphe.extension.aliexpress.patches;

import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("unused")
public final class RemoveAdsPatch {

    public static boolean shouldShowAd() {
        return false;
    }

    public static List<Object> filterAds(List<Object> items) {
        if (items == null || items.isEmpty()) return items;
        List<Object> filtered = new ArrayList<>();
        for (Object item : items) {
            if (!ItemUtils.isAdOrSponsored(item)) {
                filtered.add(item);
            }
        }
        return filtered;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/aliexpress/ads/RemoveAdsPatch.kt
package app.morphe.patches.aliexpress.ads

import app.morphe.patcher.Fingerprint
import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.aliexpress.utils.Constants.COMPATIBILITY_ALIEXPRESS

internal object AdcInitFingerprint : Fingerprint(
    returnType = "V",
    parameters = listOf("Landroid/app/Application;"),
)

val removeAdsPatch = bytecodePatch(
    name = "Remove ads",
    description = "Removes sponsored products and advertisements from the AliExpress home feed and search results.",
    default = true,
) {
    compatibleWith(COMPATIBILITY_ALIEXPRESS)
    dependencies = listOf("Bypass signature check")

    execute {
        val targetMethod = AdcInitFingerprint.methodOrNull ?: throw IllegalStateException(
            "Failed to resolve target method for AdcInitFingerprint. Ensure the fingerprint matches the current application version."
        )
        targetMethod.apply {
            addInstructions(
                0,
                """
                invoke-static {}, Lapp/morphe/extension/aliexpress/patches/RemoveAdsPatch;->shouldShowAd()Z
                move-result v0
                if-nez v0, :cond_skip
                invoke-static {p0}, Lapp/morphe/extension/aliexpress/patches/RemoveAdsPatch;->filterAds(Ljava/util/List;)Ljava/util/List;
                move-result-object p0
                :cond_skip
                """.trimIndent(),
            )
        }
    }
}
